http2: refactor how trailers are done

jasnell opened the pull request:
Rather than an option, introduce a method and an event...

```js
server.on('stream', (stream) => {
  stream.respond(undefined, { hasTrailers: true });
  stream.on('trailers-ready', () => {
    stream.trailers({ abc: 'xyz'});
  });
  stream.end('hello world');
});
```

The `'trailers-ready'` event is necessary because of
`respondWithFile` and `respondWithFD`... neither of
which cause the `Http2Stream` to emit the normal stream
related events (by design).

If using the Stream API, however, we can also do:

```js
server.on('stream', (stream) => {
  stream.respond(undefined, { hasTrailers: true });
  stream.on('finish', () => {
    stream.trailers({ abc: 'xyz' });
  });
  stream.end('hello world');
```

This is a breaking change in the API such that the prior
`options.getTrailers` is no longer supported at all.
Ordinarily this would be semver-major and require a
deprecation but the http2 stuff is still experimental.

/cc @mcollina @addaleax @sebdeckers @apapirovski @trivikr @nodejs/http2

##### Checklist
- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

hiroppy commented on doc/api/http2.md:
```diff
@@ -1420,29 +1445,25 @@ The `options.onError` function may also be used to handle all the errors
-if the `getTrailers` callback attempts to set such header fields.
-
-### Class: Http2Server
+s### Class: Http2Server
```
`s`

jasnell replied:
heh... silly typo

vsemozhetbyt commented on doc/api/http2.md:
```diff
@@ -694,8 +694,8 @@ added: v8.4.0
     and `256` (inclusive).
-  * `getTrailers` {Function} Callback function invoked to collect trailer
-    headers.
+  * `hasTrailers` {boolean} When `true`, the `Http2Stream` will emit the
```
Do we update YAML metadata when we add an options field? If yes, there are 4 such additions in this PR.

mcollina replied:
I would rather call this `waitForTrailers`.

vsemozhetbyt commented on doc/api/http2.md:
```diff
@@ -1036,6 +1042,35 @@ Provides miscellaneous information about the current state of the
+
+Sends a trailing `HEADERS` frame to the connected HTTP/2 peer. This method
+will cause the `Http2Stream` to be immediately closed and should only be
+called when the last DATA frame has been sent to the peer. When sending a
```
I am not sure if this is a subject for this PR, but there is some inconsistency in frame names (HEADERS, DATA) — sometimes they are wrapped in backticks, sometimes not.

vsemozhetbyt commented on doc/api/http2.md:
```diff
@@ -1036,6 +1042,35 @@ Provides miscellaneous information about the current state of the
 
 A current state of this `Http2Stream`.
 
+#### http2stream.trailers(headers)
```
Are we OK with using non-verbs for method names? There are some other examples in http2 API, so this is just to be sure that this is intended)

mcollina replied:
maybe this can be `sendTrailers()`?

apapirovski commented:
I assume the main reason here is that this enables async behaviour while gather the trailer headers, as opposed to the old API which forced sync behaviour?

I've only had a chance to skim but what would be the effort level to always hang it off `finish` event rather than introducing another custom event?

mcollina commented on lib/internal/http2/core.js:
```diff
@@ -1745,6 +1728,31 @@ class Http2Stream extends Duplex {
+    if (this.destroyed || this.closed)
+      throw new ERR_HTTP2_INVALID_STREAM();
+    if (this[kSentTrailers])
+      throw new ERR_HTTP2_TRAILERS_ALREADY_SENT();
```
Can you please add a test for this case?

mcollina commented on doc/api/http2.md:
```diff
@@ -1296,10 +1326,10 @@ Note that using the same file descriptor concurrently for multiple streams
+When the `options.hasTrailers` option is set, the `'trailers-ready'` event will
+be emitted immediately after queuing the last chunk of payload data to be sent.
+The `http2stream.trailers()` method can then be used to sent trailing header
+fields to the peer.
```
if `hasTrailers: true`, then the user __must__ call `.trailers()` to close the writable side of the stream. This is not being stressed enough in the sentence. As far as I understand, if `.trailers()`  is not called,  there will be a memory leak.
This seems a dangerous API.

mcollina replied:
All of that should definitely be documented.

mcollina commented on lib/internal/http2/core.js:
```diff
@@ -1745,6 +1728,31 @@ class Http2Stream extends Duplex {
     priorityFn();
   }
 
+  trailers(headers) {
```
what will happen if `trailers(headers)` is called _before_ `end()`? This case should be tested.

mcollina replied:
This is very dangerous. I would have `trailers` throw if the event has not be emitted yet instead.

mcollina commented on doc/api/http2.md:
````diff
@@ -874,6 +870,16 @@ stream.on('trailers', (headers, flags) => {
 });
 ```
 
+#### Event: 'trailers-ready'
````
Considering that the other side of the duplex can have trailers as well, maybe we need a better name for this, or maybe just emit `'finish'` in the cases it's a `respondWithFile()`

jasnell replied:
We *could* but when using `respondWithFile()`, one of the things that happens automatically is that `end()` is called on the stream API interface to ensure that someone does not attempt to write to the `Duplex` after calling `respondWithFile()`... I believe that will have the result of causing the stream to emit `'finish'` as part of the stream API processing. I'd prefer to have a single event that would work for both cases so I'm open to names.

jasnell replied:
We could use the `'finish'` event if we change the way we handle closing the `Duplex` when `respondWith*` is used. Basically, don't end the Writable side but error if `write()` is used after `respondWith*` is used.

mcollina replied:
+1 for two separate things.

I'm not necessarily certain that calling `end()` is the right thing to do in that case. But that's another issue.

addaleax commented on doc/api/http2.md:
```diff
@@ -722,14 +722,10 @@ req.on('response', (headers) => {
-The HTTP/1 specification forbids trailers from containing HTTP/2 pseudo-header
-fields (e.g. `':method'`, `':path'`, etc). An `'error'` event will be emitted
-if the `getTrailers` callback attempts to set such header fields.
+When the `options.hasTrailers` option is set, the `'trailers-ready'` event
```
Our docs currently list 21 event names that consist of more than one word; All of them use `lowerCamelCase`, so it might be good to be consistent with that?

addaleax commented on lib/internal/http2/core.js:
```diff
@@ -295,25 +296,14 @@ function tryClose(fd) {
-  stream[kSentTrailers] = trailers;
-  return headersList;
+    return;
+  process.nextTick(emit, stream, 'trailers-ready');
```
What happens if the event listener never calls `.trailers()`?

jasnell replied:
Then it doesn't close and things hang until it does. Exactly the same as if `end()` is never called on an HTTP `response` object.

The stream can be closed using either `trailers()` or `close()` on the `Http2Stream`.

That should definitely be emphasized more in the docs.

addaleax replied:
Okay … I don’t see any way to do that better, I guess, but could we at least make it an early error if there are no listeners for the event at this point?

jasnell replied:
potentially, yes, but doing so would lock us in to having those trailers available right when that event is triggered when it could be any time after that event. Will be stewing on this more.

(this, btw, is the part I've been stewing on for a while with this which is why it took so long to get this done. I haven't been able to come up with a better solution.)

addaleax replied:
I meant basically adding `if (stream.listenerCount('trailers-ready') === 0) throw new Error();` here, right before this line – nothing fancier than that

jasnell replied:
Or just simply `close()` rather than throw. That's certainly possible. The `trailers-ready` (or whatever we use) is a marker that says trailers can be sent at any time after this. If the user is not listening for that, then they cannot reasonably know when to proceed with sending the trailers anyway.

jasnell commented:
Ok, I made several updates

1. I renamed the `hasTrailers` option to `waitForTrailers` as requested
2. I renamed the `trailers()` function to `sendTrailers()` as requested
3. I renamed the `trailers-ready` event to `wantTrailers` ... using `finish` here includes a variety of complexities with the `Writable` API.. see https://github.com/nodejs/node/issues/20014
4. I stressed the need to call `close()` or `sendTrailers()` in the docs to properly close the `Http2Stream` when using `wantTrailers`
5. Fixed a number of nits

jasnell commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/14274/

jasnell commented:
@mcollina @addaleax @nodejs/http2 .. when you get a moment, I'd appreciate another quick review. I'd like to get this landed early next week in time for the next 10.0.0 test build.

jasnell commented:
Landed in 237aa7e
